util/reggen/validate.py

```diff
@@ -1151,115 +1157,104 @@ def validate_multi(mreg, offset, addrsep, width, top):
-            genreg['desc'] = mreg['desc']
-            genreg['hwext'] = mreg['hwext']
-            genreg['hwqe'] = mreg['hwqe']
-            genreg['hwre'] = mreg['hwre']
-            genreg['swaccess'] = default_sw
-            genreg['hwaccess'] = default_hw
-            if 'tags' in mreg:
-                genreg['tags'] = mreg['tags']
-            else:
-                genreg['tags'] = []
+    idx = 0
 
-            if regwen_incr:
-                genreg['regwen'] = mreg['regwen'] + str(rnum)
-            else:
-                genreg['regwen'] = mreg['regwen']
-            genreg['shadowed'] = mreg['shadowed']
-
-            resval = 0
-            resmask = 0
-            bits_used = 0
-            genfields = []
-
-        while bpos < width:
-            # bused is a bit mask of how many bits the fields need
-            # bpos is the current LSB of the bits allocated for the fields
-            trypos = bused << bpos
-
-            # if register can no longer contain another homogenous field
-            # if the field is not homogenous and has been allocated
-            # Currently we only compact if field is homogenous
-            if (trypos > max_rval) or (bpos != 0 and len(mreg['fields']) > 1):
-                bpos = width
-                break
-            # if bits have not been used, break and allocate
-            if (trypos & bits_used) == 0:
-                break
-            bpos += 1
-
-        if bpos < width:
-            # found a spot
-            for fn in mreg['fields']:
-                newf = fn.copy()
-                newf['name'] += "_" + str(inum)
-                if bpos != 0:
-                    newf['bits'] = bitfield_add(newf['bits'], bpos)
-                    newf['desc'] = 'for ' + cname + str(inum)
-                    newf['bitinfo'] = (newf['bitinfo'][0] << bpos,
-                                       newf['bitinfo'][1],
-                                       newf['bitinfo'][2] + bpos)
-                    if 'tags' not in fn:
-                        newf['tags'] = []
-                    if 'enum' in newf:
-                        del newf['enum']
-                else:
-                    newf['desc'] += ' for ' + cname + str(inum)
-                genfields.append(newf)
-            bits_used = bits_used | bused << bpos
-            resval = resval | gen[1] << bpos
-            resmask = resmask | gen[2] << bpos
-            bpos += 1
-            inum += 1
-            if inum == mcount:
-                closereg = True
+    log.info("Multireg attributes 0x{:x} {} {} {}".format(field_bitmask, bits_used,
+                                                          num_fields, max_fields_per_reg))
+
+    while idx < mcount:
+
+        genreg = deepcopy(template_reg)
+
+        # will fields be compacted?
+        is_compact = num_fields == 1 and genreg['compact']
+
+        # will there be multiple registers?
+        is_mult = (mcount > max_fields_per_reg) or (not is_compact and mcount > 1)
```

Comment:
`genreg['compact']` is always set to `mreg['compact']`, right? would it make sense to move the `is_compact` and `is_mult` out of the loop?

Reply:
yes good point i have moved it out.